whisper.cpp

```diff
@@ -4293,10 +4303,19 @@ static void whisper_process_logits(
             //log("timestamp_logprob=%f max_text_token_logprob=%f\n", timestamp_logprob, max_text_token_logprob);
 
             if (timestamp_logprob > max_text_token_logprob) {
+                //printf("sampling timestamp\n");
                 for (int i = 0; i < vocab.token_beg; ++i) {
                     logits[i]   = -INFINITY;
                     logprobs[i] = -INFINITY;
                 }
+            } else {
+                //printf("sampling text\n");
+                for (int i = vocab.token_beg; i < n_logits; ++i) {
+                    logits[i]   = -INFINITY;
+                    logprobs[i] = -INFINITY;
+                }
+
+                whisper_suppress_invalid_grammar(ctx, params, logprobs, decoder.grammar);
```

Comment:
Moving this down here improves the results in my tests significantly. The idea is that we keep the logic for sampling a timestamp token intact, irrespective of what the grammar suggests. Only after we have made the decision about the next token being text, we apply the grammar filters.

Reply:
Makes sense! This was my first foray into whisper internals so I don't have a nuanced understanding of the decoding process.